drivers/clock_control/clock_control_mcux_ccm_rev3.c

```diff
@@ -0,0 +1,555 @@
+}
+
+static int _mcux_ccm_on(const struct device *dev,
+			struct imx_ccm_clock *clk)
+{
+	int ret;
+
+	LOG_DBG("currently ungating clock %s", clk->name);
+
+	if (!clk->parent) {
+		goto out_ungate;
+	}
+
+	ret = _mcux_ccm_on(dev, clk->parent);
+	if (ret < 0) {
+		LOG_ERR("failed ungating operation for clock %s", clk->parent->name);
+		return ret;
+	}
+
+out_ungate:
+	return mcux_ccm_on_off(dev, clk, true);
+}
+
+static int mcux_ccm_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	/* ungate current clock and its parents */
+	return _mcux_ccm_on(dev, clk);
+}
+
+static int mcux_ccm_off(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	return mcux_ccm_on_off(dev, clk, false);
+}
+
+static int mcux_ccm_get_rate(const struct device *dev,
+			     clock_control_subsys_t sys, uint32_t *rate)
+{
+	const struct imx_ccm_config *cfg;
+	struct imx_ccm_data *data;
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	cfg = dev->config;
+	data = dev->data;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	/* clock not configured yet */
+	if (!clk->freq) {
+		LOG_ERR("can't get rate of unconfigured clock %s: %d",
+			clk->name, ret);
+		return -EINVAL;
+	}
+
+	*rate = clk->freq;
+
+	return 0;
+}
+
+static int _mcux_ccm_set_rate(const struct device *dev,
+			      struct imx_ccm_clock *clk,
+			      uint32_t rate)
+{
+	uint32_t parent_rate;
+	int ret, obtained_rate, clk_state;
+
+	clk_state = clk->state;
+
+	LOG_DBG("trying to set rate %u for clock %s", rate, clk->name);
+
+	/* note: although a set_clock_rate() operation may not
+	 * yield a frequency equal to the requested rate, this
+	 * will help filter out the cases in which it does.
+	 *
+	 */
+	if (clk->freq == rate) {
+		LOG_ERR("clock %s already set to rate %u", clk->name, rate);
+		return -EALREADY;
+	}
+
+	/* can't go any further in the clock tree */
+	if (!clk->parent) {
+		goto out_set_rate;
+	}
+
+	ret = imx_ccm_get_parent_rate(dev, clk, clk->parent, rate, &parent_rate);
+	if (ret == -EPERM || ret == -EALREADY) {
+		LOG_DBG("early stop in tree traversal for clock %s", clk->name);
+		/* we're not allowed to go up the clock hierarchy */
+		goto out_set_rate;
+	} else if (ret < 0) {
+		LOG_ERR("failed to get parent rate for clock %s: %d",
+			clk->name, ret);
+		return ret;
+	}
+
+	/* go up the clock hierarcy in order to set the parent's rate */
+	ret = _mcux_ccm_set_rate(dev, clk->parent, parent_rate);
```

Comment:
I'm unclear why we go to the clock parent to set the rate- will this cause dependency problems if the parent clock has other child clock dependencies?